Get bookmarks with collections, date-times and comment threads

bianchi-dy opened the pull request:
**This is a ...**
New feature (non-breaking change which adds functionality)

**What does this PR do?**
Scrapes the following information:

**Author username**
Users who have bookmarked a work, with tags, collections, date, and time
Comment threads (even the nested ones!) with users, date, and time

**How was the change made?**
Added new functions and columns per item, i.e. `get_bookmarks, get_all_comments`. `get_all_comments` is nested as a sort of hack since it uses recursion to get to the deepest level of the comment thread.

**Things I'd like feedback on**
Code elegance, I suppose?

**What next?**
- [ ] Fix conflicts
- [ ] Check if everything is working with the new merges 
- Working on a fandom analysis study with some friends from the YOI fandom! I'd like to credit this repository if we end up publishing anything, if that would be ok with you two.

ssterman committed:
Update README.md

move --lang flag to correct example call

AJHoeh committed:
Fix #20

AJHoeh committed:
Fix #22

AJHoeh committed:
Fix #23

AJHoeh committed:
Fix #23 for ao3_get_fanfics.py

ssterman committed:
fix soup selectors for multiple classes

ssterman committed:
Merge branch 'master' of https://github.com/radiolarian/AO3Scraper

ssterman committed:
make bookmarks optional with a --bookmarks flag; add lang handling to restart and single fic conditions; add the page delay to the bookmark condition

AJHoeh committed:
Merged fix for #23

AJHoeh committed:
Implemented 429 status handling in ao3_work_ids.py

AJHoeh committed:
Implemented status code handling for ao3_get_fanfics.py

celegant commented:
Hi there! There seems to be something wrong with scraping works with only a single page of kudos, as the URLs don't have this requirement: soup.find(id='kudos_more_link')['href'] and it leads to a "TypeError: 'NoneType' object is not subscriptable" error.

bianchi-dy commented:
Hi @celegant I think that's AO3 changing its source code again. I might have time in the next few weeks to fix that and push something up

Vicky-Vicky27 committed:
Updated Dependependencies

mrolfe committed:
changed is to == or !=

compatible with python 3.9

Commit:
allow output to be in a different directory and handle error files okay

Commit:
allow option to only fetch metadata; skips pulling content

Commit:
add --metadata-only flag info to the README

Commit:
Load existing ids from CSV file

* reads the given CSV file for output and reads in the existing IDs.
  * prevents duplicates of IDs when rerunning the command
* also makes seen_ids a set for better lookup.

Commit:
Retry ao3_work_ids request on 429 after 5s delay

Previously, script would quit having decided there are no works on the
429 page. This is annoying if you've got some hundreds of pages you're
going through. The script quit on me after about ~100-150 pages.

ssterman committed:
Merge pull request #32 from mrolfe/fixed_is

changed is to == or !=

ssterman committed:
Merge pull request #26 from AJHoeh/master

Fix for #23

ssterman committed:
Merge pull request #27 from AJHoeh/fix_#24

Fix for #24

ssterman committed:
Merge pull request #31 from Vicky-Vicky27/README-Update

Updated Dependencies

ssterman committed:
Merge pull request #34 from katfang/fetch-metadata-only

adds --metadata-only option to skip pulling content

ssterman committed:
update description of metadata_only

ssterman committed:
Merge branch 'master' into output-directory-fix

ssterman committed:
Merge pull request #35 from katfang/output-directory-fix

Allow output csv when fetching fics to be in a different directory and handle error files appropriately

ssterman committed:
Update ao3_get_fanfics.py

Create a directory if the pathname for the output file includes a new directory.

ssterman committed:
Merge branch 'master' into load-existing-seen-ids

ssterman committed:
Merge pull request #37 from katfang/load-existing-seen-ids

ao3_work_ids improvements -- load existing seen_ids, retry on 429

ssterman committed:
skip loading seen ids if no file yet exists

ssterman committed:
don't try to make a new directory if the output csv is only a filename
